### ISSUE TITLE
Question about the propagation and over-smoothing

### ISSUE THREAD

**Issue by bjarkedc** [roth-andreas/simplifying-over-smoothing#1]
Hey, I really like your work on rank-collapse, and I am trying to understand the way the Dirichlet energy and rank_diff are calculated for each layer for each GNN.

From looking at the code, I understand that there is no actual optimization, i.e., backpropagation done. We just propagate through with each convolution, by message passing (aggregate and update), and at each layer, we have the hidden representation of the features for the nodes in the graph. Is this understood correctly?

Does this mean that when showing that a method such as GCN over-smooths or suffers from rank-collapse, we only look at it through the propagation without optimization? If so, do you know if this is the same approach that has been used in ["A Survey On Oversmoothing In Graph Neural Networks"](https://arxiv.org/abs/2303.10993) 

Is this the only way to show these things empirically, or can you also show that a model over-smooths after you try to train it? For example, a scenario where I train a model with 32 layers and then propagate my data through to fetch the representation of the nodes for each layer, and then finally calculate the rank_diff and Dirichlet energy for each layer. Would this prove it over-smooths if they both decay to low numbers?

Sorry, this is a lot of questions, and I'm kinda new to this, so feel free to delete or close. Thank you in advance.

**Comment by roth-andreas** [comment]
Great to see your interest in this work!

You are correct. All calculations for this work are calculated using the randomly initialized models without optimization. This is also done in ["A Survey On Oversmoothing In Graph Neural Networks"](https://arxiv.org/abs/2303.10993) and all other studies on over-smoothing that I am aware of. 

Concerning optimization (consider the GCN for simplicity): The theory on over-smoothing indicates that there exist parameters for which over-smoothing can be prevented at any depth (see ["[Rank Collapse Causes Over-Smoothing and Over-Correlation in Graph Neural Networks](https://proceedings.mlr.press/v231/roth24a.html) for more details). However, there will be fewer and fewer valid parameter matrices with increased depth (the volume of their space goes to zero; convergence in probability). These cannot be found by gradient descent, so with large enough depth training the model will not change anything. However, when the Dirichlet energy and the rank_diff are from zero, gradient descent can find better parameter matrices.

Feel free to ask further questions.

**Comment by bjarkedc** [comment]
Thank you so much for your quick reply! 

Great example and explanation, thank you! 

I have so many questions, but i will try to keep it to this one

In your paper [Rank Collapse Causes Over-Smoothing and Over-Correlation in Graph Neural Networks](https://proceedings.mlr.press/v231/roth24a.html). You make the connection between a GCN (what i understand as Eq.1) and then finally you show that a sum of Kronecker products would be a solution and then show this empirically. In you next paper [Simplifying the Theory on Over-Smoothing](https://arxiv.org/pdf/2407.11876) you suggest  in part 4 after definition 4.4 that "...to prevent over-smoothing and rank collapse. It needs to be ensured that the graph convolution is not a Kronecker product, i.e., a single aggregation and transformation matrix. One direction is to operate on multiple computational graphs A1, . . . , Al with distinct feature transformations W1, . . . , Wl" and then present equation 14.

Is there any examples of any Graph Neural Network architecture that resembles the behaviour like equation 14? I was wondering if heterogeneous message passing networks like RGCN is an example of this? Or does all this strictly hold for homogeneous GNNs like Gated GNNs and Residual GNN examples? 

I apologize if this makes no sense and I just misunderstand everything.

**Comment by roth-andreas** [comment]
That is exactly right. Heterogeneous GNN can be seen as operating on multiple graphs (one for each edge type) and would, therefore, not over-smooth. There is not much out there for homogeneous GNNs in that direction. We are about to upload a paper that digs deeper by showing some nice theoretical conditions that are required for operating on multiple graphs. We will propose some instances of this framework in there as well. Hopefully, this will make it a lot clearer, and more people will start to see the advantages of that direction.

**Comment by bjarkedc** [comment]
Thank you so much for the answer. Is there a place where i can subscribe to in order to get and update on when you release it?

**Comment by bjarkedc** [comment]
Okay i know im pushing my luck here. But can you explain your way of calculating the dirichlet energy? What does the "rw" parameter stand for? row wise? random walk?

Again thank you!

**Comment by roth-andreas** [comment]
Hi @bjarkedc, my apologies for the long pause. We recently uploaded a preprint for splitting a graph A into multiple computational graphs A_1,...,A_k. It might clarify some things: https://arxiv.org/abs/2409.11504

Feel free to ask further questions!
